Automattic/wp-calypso
# Jetpack Connect: abstract an HoC that handles connection redirects for JP sites.

#### Changes proposed in this Pull Request
* move the Jetpack connection functionality to an HoC avoid code duplication.

This is a preparation for the WP.com/JP [Jetpack Search purchase integration
](https://github.com/Automattic/wp-calypso/pull/41748) to handle non-Jetpack sites and move away from the `jetpack/connect` route. Purchasing a product for the Jetpack sites includes connection checks, shared with the Jetpack Connect Flow.  

The change is substantial, so for ease of review, this PR avoids major re-writes e.g. introducing Redux. Any further improvements can be handled separately. 

#### Testing instructions

Connection Flow should change for none of the Jetpack statuses.

* For any Jetpack plan, start the connection flow, e.g. `/jetpack/connect/premium` and verify that form copy, footer, input box, redirects are correct and functional, and that the flow leads to checkout.  In order to verify that the plugin was correctly setup, to be tested with a Jetpack site that is unconnected, plugin deleted, or inactive.
* the flow starting at `/jetpack/connect` without plan specified should end at the canonical plans page.
